### PR TITLE
Fix ``unsupported-binary-operation`` on classes that overload or

### LINKED ISSUE #4951: DjangoRestFramework: E1131: unsupported operand type(s) for | (unsupported-binary-operation)
### Bug description

```python
# pylint: disable=missing-docstring

from rest_framework.permissions import BasePermission, IsAuthenticated, SAFE_METHODS

class ReadOnly(BasePermission):
    def has_permission(self, request, view):
        return request.method in SAFE_METHODS

permission_classes = [IsAuthenticated|ReadOnly]
```

### Configuration

_No response_

### Command used

```shell
pylint binary.py
```

### Pylint output

```shell
************* Module binary
binary.py:9:22: E1131: unsupported operand type(s) for | (unsupported-binary-operation)

-----------------------------------
Your code has been rated at 0.00/10
```

### Expected behavior

For pylint to not throw this error. 

Django Rest Framework supports binary operands for permissions [1]
This example code is taken from the documentation [2]
[1] https://github.com/encode/django-rest-framework/blob/master/rest_framework/permissions.py
[2] https://www.django-rest-framework.org/api-guide/permissions/

### Pylint version

```shell
pylint 2.10.2
astroid 2.7.3
Python 3.9.4 (default, Apr 10 2021,
```

### OS / Environment

Debian GNU/Linux 10 (buster)

### Additional dependencies

djangorestframework==3.12.4

*edit*: Added link [2] to django rest frameworks documentation on permissions

### PR DESCRIPTION
**Pull request by timmartin**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

We have a specific checker looking for people using `|` with two types, apparently on the assumption that anyone doing that is mistakenly trying to use union syntax where it's not supported. This seems to overlook the fact that it is possible for class definitions to use a metaclass that implements `__or__` or `__ror__` as appropriate, which is the case in the permission classes for Django REST Framework, for example.

Closes #4951

### CODE REVIEW

File: pylint/checkers/typecheck.py
Comment (timmartin) [reply pylint/checkers/typecheck.py:1920]: If you mean Python built-in `hasattr` then I don't think so, because we need to look up the attribute in Astroid's inferred knowledge of the real class, not in the `left_obj` runtime object. If there's a Pylint equivalent to `hasattr` on inferred representations then we could use that, but I couldn't find one.

Comment (Pierre-Sassoulas) [reply pylint/checkers/typecheck.py:1920]: Do you think it's worth it to add hasattr in astroid ?

Comment (timmartin) [reply pylint/checkers/typecheck.py:1920]: Sure, I'd be happy to give it a go.

**Comment by timmartin** [comment]
I'm a bit confused about the functional tests. It looks like with my changes in place, we stop reporting this warning when run under Python 3.10 (because the `type` metaclass now defines `__or__`, and Astroid picks that up) but that's not how it's supposed to work.

If we're running on Python 3.10 but the `.rc` file sets `py-version=3.8` should we report errors according to the `py-version`? If that's the case, then we're in trouble as we'll have to differentiate between a case where `__or__` is coming from `type` (in which case we should only use it if we're configured for a version of Python that defines it) or whether it's in user-defined code, in which case we can use it on any version of Python.

Is that correct?

**Comment by Pierre-Sassoulas** [comment]
> If we're running on Python 3.10 but the .rc file sets py-version=3.8 should we report errors according to the py-version?

Yes. I almost commented on this but did not because I thought the ``if not allowed_nested_syntax:`` condition might already be handling this. ``py-version`` is a valuable option but also relatively new in the codebase.

**Comment by timmartin** [comment]
It feels like the right way to solve this is for Astroid not to report the existence of an `__or__` method on Python before v3.10, because it doesn't exist. If we solve this at the Astroid level, Pylint doesn't need any special rules and Astroid will be reporting slightly more correct data. But based on a quick search of the repo, it looks like Astroid doesn't support anything like a `py-version` setting.

Does this mean we need to have logic to deal with this in the Pylint checker? If so, I think it means that we can't benefit from implementing a `hasattr`, because we'll need to inspect the attribute to see what class it belongs to.

**Comment by Pierre-Sassoulas** [comment]
> If we solve this at the Astroid level, Pylint doesn't need any special rules and Astroid will be reporting slightly more correct data. But based on a quick search of the repo, it looks like Astroid doesn't support anything like a py-version setting.

I think adding a ``py-version`` in astroid is very high effort because we'd have to know every specificity of each interpreter instead of just relying on the ast module for the interpreter (basically emulate a 3.8 parsing on a 3.10 interpreter). By contrast in pylint for a specific case we have to know what is the specificity of versions older than 3.x for this specific use case. What do you think @cdce8p ?

**Comment by timmartin** [comment]
I have what I think is a fix for this. It's not elegant, but I can see that making Astroid infer all the cases correctly based on the configured version is not going to be easy.

File: doc/whatsnew/2/2.14/summary.rst
Comment (Pierre-Sassoulas) [inline doc/whatsnew/2/2.14/summary.rst:263]: Unfortunately we need to still add the changelog in full.rst too until we release 2.14.0.

**Review by Pierre-Sassoulas — APPROVED** [review]
There's a small changelog issue (sorry about that, we're trying to get rid of the duplication in the next versions), but other than that LGTM. I'll wait for a review from someone else.

File: tests/functional/a/alternative/alternative_union_syntax_error.py
Comment (jacobtylerwalls) [inline tests/functional/a/alternative/alternative_union_syntax_error.py:112]: Sorry to suggest annoying test cases, but I've seen people do the darndest things ⚡ 
```
  File "/Users/.../pylint/pylint/checkers/typecheck.py", line 1923, in <genexpr>
    attr.parent.qname() == "builtins.type" for attr in attrs
AttributeError: 'Assign' object has no attribute 'qname'
```
```suggestion
class_list_reversed_valid = [DefaultMetaclass | WithReverse]


# Pathological cases
class HorribleMetaclass(type):
    __or__ = lambda x: x

class WithHorrible(metaclass=HorribleMetaclass):
    pass

class_list = [WithHorrible | DefaultMetaclass]
```

File: pylint/checkers/typecheck.py
Comment (jacobtylerwalls) [inline pylint/checkers/typecheck.py:1922]: You can just check the `pytype()` of the object. This might help you skirt my nasty test case.
```python
>>> import astroid
>>> l = astroid.extract_node("list")
>>> inferred = next(l.infer())
>>> inferred.pytype()
'builtins.type'
```

**Review by jacobtylerwalls — CHANGES_REQUESTED** [review]

**Comment by Pierre-Sassoulas** [comment]
Moving to 2.15 as we try to release asap :)

File: pylint/checkers/typecheck.py
Comment (timmartin) [reply pylint/checkers/typecheck.py:1922]: Thanks, it's good to know that exists. However, it doesn't seem to do what I need here. On a `nodes.ClassDef`, the `pytype()` method will always return `"builtins.type"` so I can't use it to differentiate between a user-defined class an a builtin.

**Comment by timmartin** [comment]
I don't fully understand, but on Python 3.10 Astroid will infer two values for the attribute built-in on `type`, a function definition and a bound method. However, it seems simple enough to filter out both of these.

File: doc/whatsnew/2/2.14/summary.rst
Comment (timmartin) [reply doc/whatsnew/2/2.14/summary.rst]: Now I've missed the 2.14 cut-off I guess I have to move this. I've put the note in `index.rst` in the 2.15 section because that's where all the section headings are, but I don't know if this is going to get moved into separate files later?

**Comment by timmartin** [comment]
I think you're right in your expectation that it should not report a warning with the second-level metaclass. I was assuming Astroid `getattr` would do the right thing here, but I guess it's more complicated than that. I'll have a closer look at this.

**Comment by jacobtylerwalls** [comment]
I think it might just be the assumption of direct parentage. Any time I see `.parent` I always go to try to violate that assumption somehow :D

File: tests/functional/a/alternative/alternative_union_syntax_error.py
Comment (cdce8p) [inline tests/functional/a/alternative/alternative_union_syntax_error.py:111]: Can you add this testcase also to `alternative_union_syntax.py` which sets `py-version=3.10`?

File: pylint/checkers/typecheck.py
Comment (cdce8p) [inline pylint/checkers/typecheck.py:1936]:
```suggestion
            except astroid.exceptions.NotFoundError:
```

File: pylint/checkers/typecheck.py
Comment (cdce8p) [inline pylint/checkers/typecheck.py:1945]:
```suggestion
            except astroid.exceptions.NotFoundError:
```

**Review by cdce8p — CHANGES_REQUESTED** [review]
> > If we solve this at the Astroid level, Pylint doesn't need any special rules and Astroid will be reporting slightly more correct data. But based on a quick search of the repo, it looks like Astroid doesn't support anything like a py-version setting.
> 
> I think adding a `py-version` in astroid is very high effort because we'd have to know every specificity of each interpreter instead of just relying on the ast module for the interpreter (basically emulate a 3.8 parsing on a 3.10 interpreter). By contrast in pylint for a specific case we have to know what is the specificity of versions older than 3.x for this specific use case. What do you think @cdce8p ?

It would be quite difficult to implement `py-version` in astroid unfortunately. True, we could do the AST parsing with a different version but that doesn't really change much. For instance, I believe the `__or__` and `__ror__` methods on the `type` metaclass are read directly from the interpreter. Thus we would need to manually update each and every change from Python at which point we're basically rebuilding CPython itself.

`py-version` was more designed for things which can easily be inferred by pylint, like recommendations for the use of new features (`f-strings`, ...).

The way you solved the issue here is quite clever! I left a few notes but after that I think it's ready.

File: pylint/checkers/typecheck.py
Comment (cdce8p) [inline pylint/checkers/typecheck.py:1910]:
```suggestion
        pre-3.10 compatibility then Astroid will have inferred the
```

File: pylint/checkers/typecheck.py
Comment (cdce8p) [inline pylint/checkers/typecheck.py:1919]:
```suggestion
        return not is_py310_builtin or self._py310_plus
```
The brackets are redundant. `not` binds stronger than `or`.

File: pylint/checkers/typecheck.py
Comment (cdce8p) [inline pylint/checkers/typecheck.py:1905]:
```suggestion
    def _includes_version_compatible_overload(self, attrs: list) -> bool:
```
I'm not sure about the type for `attrs`. At the very least, it can also be `astroid.bases.BoundMethod` which doesn't inherit from `nodes.NodeNG`. Maybe lets leave it at just `list` for now.

File: pylint/checkers/typecheck.py
Comment (cdce8p) [inline pylint/checkers/typecheck.py:23]:
```suggestion

import astroid
import astroid.exceptions
```
Sorry for the confusion here. I missed that the bare `astroid` import is used as well.

File: tests/functional/a/alternative/alternative_union_syntax_error.py
Comment (cdce8p) [inline tests/functional/a/alternative/alternative_union_syntax_error.py:130]: This doesn't seem to work at the moment. I only spend a few minutes investigating it, but couldn't  a quick fix. Maybe it's better to remove that test case here and open a followup issue for it to be addressed later. At least that way this PR could be merged.

File: pylint/checkers/typecheck.py
Comment (cdce8p) [reply pylint/checkers/typecheck.py:1940]: Please address the comment above

File: pylint/checkers/typecheck.py
Comment (cdce8p) [reply pylint/checkers/typecheck.py:1949]: Please address the comment above

**Review by cdce8p — CHANGES_REQUESTED** [review]

File: tests/functional/a/alternative/alternative_union_syntax_error.py
Comment (timmartin) [reply tests/functional/a/alternative/alternative_union_syntax_error.py:130]: I was hoping to get to the bottom of it, because I didn't want to commit a half-working fix. But thinking about it, this failure case isn't a new false positive so I guess this fix will be an improvement overall. I'll pull this out and open another PR for it.

**Comment by jacobtylerwalls** [comment]
Yep, thanks for checking! I was the one that asked for that test case we haven't solved, but I agree we can postpone. I think it's just a matter of doing a recursive search on ancestors instead of relying on direct `.parent`-age, but we can postpone.

**Review by jacobtylerwalls — APPROVED** [review]

**Pull request merged**
